File: core/src/main/java/com/serious/service/AbstractComponent.java
package com.serious.service;
/*
 * @COPYRIGHT (C) 2023 Andreas Ernst
 *
 * All rights reserved
 */

import lombok.extern.slf4j.Slf4j;
import org.springframework.cloud.commons.util.InetUtils;
import org.springframework.cloud.commons.util.InetUtilsProperties;

/**
 * @author Andreas Ernst
 */
@Slf4j
public abstract class AbstractComponent implements Component {
    // static data

    static String port;

    static InetUtils inet = new InetUtils(new InetUtilsProperties());

    // static methods

    public static String getHost() {
        return inet.findFirstNonLoopbackHostInfo().getHostname();// TODO getIpAddress();
    }

    public static String getPort() {
        return port;
    }

    // instance data

    private ComponentStatus status = ComponentStatus.VIRGIN;

    // implement Component
    public void startup() {
        log.info("starting up {}", getClass().getName());

        status = ComponentStatus.RUNNING;
    }

    public void shutdown() {
        log.info("shutdown {}", getClass().getName());

        status = ComponentStatus.STOPPED;
    }

    public ComponentStatus getStatus() {
        return status;
    }

    public ComponentHealth getHealth() {
        return getStatus() == ComponentStatus.RUNNING ? ComponentHealth.UP : ComponentHealth.DOWN;
    }
}

File: core/src/main/java/com/serious/service/AbstractComponent.kt
package com.serious.service
/*
* @COPYRIGHT (C) 2023 Andreas Ernst
*
* All rights reserved
*/

import com.serious.service.ServiceManager.Companion.log
import lombok.extern.slf4j.Slf4j
import org.springframework.cloud.commons.util.InetUtils
import org.springframework.cloud.commons.util.InetUtilsProperties

/**
 * abstract base class for components
 */
@Slf4j
abstract class AbstractComponent : Component {
    // implement Component

     override var status = ComponentStatus.VIRGIN

    override fun startup() {
        log.info("starting up {}", javaClass.getName())

        status = ComponentStatus.RUNNING
    }

    override fun shutdown() {
        log.info("shutdown {}", javaClass.getName())

        status = ComponentStatus.STOPPED
    }

    override val health: ComponentHealth
        get() = if (status == ComponentStatus.RUNNING) ComponentHealth.UP else ComponentHealth.DOWN

    companion object {
        // static data

        @JvmField
        var port: String? = null
        var inet = InetUtils(InetUtilsProperties())
        @JvmStatic
        val host: String
            get() = inet.findFirstNonLoopbackHostInfo().hostname // isn't getIpAddress() better?
    }
}
